### ISSUE TITLE
Test plan for "Unions"

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#81074]
## Unions Test Plan

Championed proposal: https://github.com/dotnet/csharplang/issues/9662
Speclet: https://github.com/dotnet/csharplang/blob/main/proposals/unions.md
Feature branch: TBD

**Comment by RikkiGibson** [comment]
I was poking at cases where a closed subtype gets 'included' in a type set via multiple paths, and, thought of this scenario, which should probably have a test. It seems like a warning should be reported in this scenario, but, it also seems low priority.

```cs
    [Fact]
    public void CaseTypes_IncludesBaseAndDerived_NullableDifference()
    {
        // Union including both base and derived of a hierarchy, with a nested nullability difference
        // It seems like we can't include both 'C<string>' and 'C<string?>', and, picking just one could introduce a null safety issue.
        var source = """
            #nullable enable
            using System;

            class Program
            {
                static void Main()
                {
                    try
                    {
                        M1(new D1(null));
                    }
                    catch (NullReferenceException)
                    {
                        Console.Write("caught ");
                    }

                    M2(new D1(null));
                    M1(new C<string>("abc"));
                    M2(new C<string>("abc"));
                }

                static void M1(U u)
                {
                    Console.Write(u switch
                    {
                        C<string> c => c.Value.Length,
                        _ => 2,
                    });
                    Console.Write(' ');
                }

                static void M2(U u)
                {
                    Console.Write(u switch
                    {
                        C<string?> c => c.Value?.Length ?? -1,
                        _ => 2,
                    });
                    Console.Write(' ');
                }
            }

            union U(D1, C<string>);
            class C<T>(T value) { public T Value = value; }
            class D1(string? value) : C<string?>(value) { }
            """;

        var verifier = CompileAndVerify([source, UnionAttributeSource, IUnionSource, ClosedAttributeDefinition], targetFramework: TargetFramework.Net100, expectedOutput: "caught-133");
        verifier.VerifyDiagnostics();
    }
```

Let me know if this is not the way you want this suggestion to be tracked, happy to move it to any other format.

**Comment by AlekseyTs** [comment]
@RikkiGibson 

> Let me know if this is not the way you want this suggestion to be tracked, happy to move it to any other format.

My preference would be to not convert a test plan issue to a "kitchen sink" of random discussions.

Regarding, the specific issue. It looks like pattern matching doesn't care about mismatch in nullability that might occur during a pattern matching. This can be observed in the following scenario:
```
#nullable enable

class Program
{
    static bool M1(C<string> x) => x is C<string?>;

    static bool M2(C<string?> x) => x is C<string>;

    static C<string?> M3(C<string> x) => x;
}

class C<T>;
``` 

Therefore, I don't think that nullability difference should matter for Union matching too.
